### ISSUE TITLE
Option-union arrays loose naming / typing information when all Nones.

### ISSUE THREAD

**Issue by lgray** [scikit-hep/awkward#815]
Using the following file as input, because constructing a trivial example has proven painful:
https://drive.google.com/file/d/1t4iyU103CEWDuF9Tqx36D2cPVDyxdJDP/view?usp=sharing

You will need a recent version of coffea installed to run this (just pip install coffea).
I am using the current main branch of awkward, which will soon also be in 1.30.rc2.

The issue is that when a `?union[type1 type2]` array contains all Nones it loses its ability behave as either type1 or type2, which is preserved when the array has non-None elements.

It's pretty clear when that's the only difference between consecutive 1000 event chunks in the same file.

Here's the script for reproducing the error:
```python3
from coffea import nanoevents
from coffea.nanoevents.methods import candidate
import numpy as np
import awkward as ak

nanoevents.NanoAODSchema.mixins["FatJetLS"] = "PtEtaPhiMLorentzVector"

def script_excerpt(start, stop):
    x = nanoevents.NanoEventsFactory.from_root(
        './nano106Xv8_on_mini106X_2017_mc_NANO_py_NANO_46.root',
        entry_start=start,
        entry_stop=stop,
    )

    events = x.events()

    muons = events.Muon
    electrons = events.Electron
    jets = events.Jet
    fatjets = events.FatJet
    subjets = events.SubJet
    fatjetsLS = events.FatJetLS
    met = events.MET

    goodmuon = (
        (muons.mediumId)
        & (muons.miniPFRelIso_all <= 0.2)
        & (muons.pt >= 27)
        & (abs(muons.eta) <= 2.4)
        & (abs(muons.dz) < 0.1)
        & (abs(muons.dxy) < 0.05)
        & (muons.sip3d < 4)
    )
    good_muons = muons[goodmuon]
    ngood_muons = ak.sum(goodmuon, axis=1)

    # electrons
    goodelectron = (
        (electrons.mvaFall17V2noIso_WP90)
        & (electrons.pt >= 30)
        & (abs(electrons.eta) <= 1.479)
        & (abs(electrons.dz) < 0.1)
        & (abs(electrons.dxy) < 0.05)
        & (electrons.sip3d < 4)
    )
    good_electrons = electrons[goodelectron]
    ngood_electrons = ak.sum(goodelectron, axis=1)
        
    # good leptons
    good_leptons = ak.concatenate(
        [good_muons, good_electrons],
        axis=1,
        #behavior=candidate.behavior,
    )
    print("name of good_leptons ->", good_leptons.layout.purelist_parameter("__record__"))
    print(ak.argsort(good_leptons.pt))
    good_leptons = good_leptons[ak.argsort(good_leptons.pt)]
    candidatelep = ak.firsts(good_leptons)

    # jets
    ht = ak.sum(jets[jets.pt > 30].pt,axis=1)
    goodjet = (
        (jets.isTight)
        & (jets.pt > 30)
        & (abs(jets.eta) <= 2.5)
    )
    good_jets = jets[goodjet]

    good_fatjet = (
        (fatjets.isTight)
        & (abs(fatjets.eta) <= 2.4)
        & (fatjets.pt > 50)
        & (fatjets.msoftdrop > 30)
        & (fatjets.msoftdrop < 210)
    )
    good_fatjets = fatjets[good_fatjet]

    # hbb candidate
    mask_hbb = (
        (good_fatjets.pt > 200)
        & (good_fatjets.delta_r(candidatelep) > 2.0)
    )
    candidateHbb = ak.firsts(good_fatjets[mask_hbb])

    # number of AK4 away from bb jet
    jets_awayHbb = jets[good_jets.delta_r(candidateHbb) >= 1.2]

    # fat jets Lepton Subtracted
    # TODO: add ID
    # TODO: add 2 subjets w pt > 20 & eta<2.4
    good_fatjetLS = (
        (fatjetsLS.pt > 0)
        # & (fatjetsLS.delta_r(candidatelep) > 1.2)
    )
    good_fatjetLSs = fatjetsLS[good_fatjetLS]

    # wqq candidate
    mask_hww = (
        (good_fatjetLSs.mass > 10)
    )
    candidateWjj = ak.firsts(fatjetsLS)
    print(len(candidateWjj.pt))
    print(candidateWjj.pt)
    candidateWjj = ak.firsts(
        good_fatjetLSs[mask_hww][ak.argmin(
            good_fatjetLSs[mask_hww].delta_r(candidatelep),
            axis=1,
            keepdims=True)
        ]
    )
    print(len(candidateWjj.pt))
    print(candidateWjj.pt)

    # TODO: add lvqq likelihood and HWW mass reconstruction
    # For now, reconstruct the mass by taking qq jet, lepton and MET and solving for the z component of the neutrino momentum
    # by requiring that the invariant mass of the group of objects is the Higgs mass = 125
    def getNeutrinoZ(vis,inv,h_mass=125):
        a = h_mass*h_mass - vis.mass*vis.mass + 2*vis.x*inv.x + 2*vis.y*inv.y
        A = 4*(vis.t*vis.t - vis.z*vis.z)
        B = -4*a*vis.z
        C = 4*vis.t*vis.t*(inv.x*inv.x + inv.y*inv.y) - a*a
        delta = B*B - 4*A*C
        invZ = ((delta<0)*( -B/(2*A) )
               + (delta>0)*( np.maximum( (-B + np.sqrt(delta))/(2*A), (-B - np.sqrt(delta))/(2*A)) ))
        neutrino =  ak.zip({"x": inv.x,
                            "y": inv.y,
                            "z": invZ,
                            "t": np.sqrt(inv.x*inv.x + inv.y*inv.y + invZ*invZ),
                        },
                        with_name="LorentzVector")
        return neutrino

    print(candidatelep)
    print(candidateWjj)

    vis = candidatelep + candidateWjj  # error occurs here!!!!
            
    candidateNeutrino = getNeutrinoZ(candidatelep + candidateWjj, met)

if __name__ == "__main__":
    # works
    script_excerpt(99860, 100860)
    
    # doesn't work
    script_excerpt(100860, 101878)
```

**Comment by nsmith-** [comment]
I was working on a simpler example but ended up hitting a different bug with union arrays and types:
```python
import numpy as np
import awkward as ak


@ak.mixin_class(ak.behavior)
class Point:
    @ak.mixin_class_method(np.add, {"Point"})
    def point_add(self, other):
        return ak.zip(
            {"x": self.x + other.x, "y": self.y + other.y}, with_name="Point",
        )


@ak.mixin_class(ak.behavior)
class Point2(Point):
    pass


def make(name):
    return ak.zip(
        {
            "x": ak.Array([[1, 2], [3]]),
            "y": ak.Array([[1, 2], [3]]),
        },
        with_name=name,
    )

a, b, c = make("Point"), make("Point2"), make("Point")
a + b
b + c
a + c
d = ak.concatenate([a, b], axis=1)
c + d  # breaks

d = ak.concatenate([a[a.x < 0], b[b.x < 0]], axis=1)
c + d  # may break more
```

**Comment by nsmith-** [comment]
Sorry for the noise earlier, but here is a simple example:
```python
import numpy as np
import awkward as ak


@ak.mixin_class(ak.behavior)
class Point:
    @ak.mixin_class_method(np.add, {"Point"})
    def point_add(self, other):
        return ak.zip(
            {"x": self.x + other.x, "y": self.y + other.y}, with_name="Point",
        )


@ak.mixin_class(ak.behavior)
class Point2(Point):
    pass


def make(name):
    return ak.zip(
        {
            "x": ak.Array([[1, 2], [3]]),
            "y": ak.Array([[1, 2], [3]]),
        },
        with_name=name,
    )

a = ak.zip(
    {
        "x": ak.Array([1, 1]),
        "y": ak.Array([1, 1]),
    },
    with_name="Point",
)
b, c = make("Point2"), make("Point")
a + b
b + c
a + c
d = ak.concatenate([b, c], axis=1)
a + d

e = ak.concatenate([b[b.x < 0], c[c.x < 0]], axis=1)
a + e
```

produces
```
Traceback (most recent call last):
  File "unionnone.py", line 43, in <module>
    a + e
  File "/usr/local/lib/python3.8/site-packages/numpy/lib/mixins.py", line 21, in func
    return ufunc(self, other)
  File "/Users/ncsmith/src/awkward-1.0/awkward/highlevel.py", line 1379, in __array_ufunc__
    return ak._connect._numpy.array_ufunc(ufunc, method, inputs, kwargs)
  File "/Users/ncsmith/src/awkward-1.0/awkward/_connect/_numpy.py", line 206, in array_ufunc
    out = ak._util.broadcast_and_apply(
  File "/Users/ncsmith/src/awkward-1.0/awkward/_util.py", line 1066, in broadcast_and_apply
    out = apply(broadcast_pack(inputs, isscalar), 0, user)
  File "/Users/ncsmith/src/awkward-1.0/awkward/_util.py", line 828, in apply
    outcontent = apply(nextinputs, depth + 1, user)
  File "/Users/ncsmith/src/awkward-1.0/awkward/_util.py", line 880, in apply
    outcontent = apply(nextinputs, depth + 1, user)
  File "/Users/ncsmith/src/awkward-1.0/awkward/_util.py", line 662, in apply
    return apply(nextinputs, depth, user)
  File "/Users/ncsmith/src/awkward-1.0/awkward/_util.py", line 737, in apply
    return apply(nextinputs, depth, user)
  File "/Users/ncsmith/src/awkward-1.0/awkward/_util.py", line 644, in apply
    return apply(nextinputs, depth, user)
  File "/Users/ncsmith/src/awkward-1.0/awkward/_util.py", line 958, in apply
    raise ValueError(
ValueError: cannot broadcast records in this type of operation

(https://github.com/scikit-hep/awkward-1.0/blob/1.2.0rc2/src/awkward/_util.py#L960)
```

**Comment by lgray** [comment]
Great - it's good to have the same error being emitted with different array structures with the same trigger (empty lists).

**Comment by jpivarski** [comment]
PR #816 fixes @nsmith-'s example, which is easier to reproduce, and is now [added to the tests](https://github.com/scikit-hep/awkward-1.0/blob/d11ac8f158cd0bec6d98f5e69d756e7e8e861960/tests/test_0815-broadcast-union-types-to-all-possibilities.py). (I'm going to try your example now, @lgray.)

The last line in @nsmith-'s example becomes

```python
>>> a + e
<PointArray [[], []] type='2 * var * Point["x": int64, "y": int64]'>
```

In this fix, I backed off on the "design choice" I was talking about yesterday: this is a real problem; we can't have `+` stop working. I guess broadcasting can't be a type-shrinkwrapping operation, though others can and should be. Broadcasting is somewhat invisible—arrays automatically change shape to fit the operation they're needed in—maybe that's why it should be treated differently.

[The change](https://github.com/scikit-hep/awkward-1.0/pull/816/commits/d11ac8f158cd0bec6d98f5e69d756e7e8e861960) replaces an iteration over `np.unique(combos)` with an `all_combos` that is constructed with `itertools.product`. (It's not an _O(n)_ iteration in the length of arrays _n_, but _O(t1*t2*t3* ...)_ in the number of tags _ti_ in each union involved in the broadcast.) In my non-empty test case,

```python
>>> c = ak.concatenate([
...     ak.with_parameter([1, 2, 3], "__array__", "one"),
...     ak.with_parameter([1.1, 2.2, 3.3], "__array__", "two"),
... ])
>>> c
<Array [1, 2, 3, 1.1, 2.2, 3.3] type='6 * union[int64[parameters={"__array__": "...'>
>>> c.type
6 * union[int64[parameters={"__array__": "one"}], float64[parameters={"__array__": "two"}]]
>>> ak.broadcast_arrays(c, c)
[
    <Array [1, 2, 3, 1.1, 2.2, 3.3] type='6 * union[int64[parameters={"__array__": "...'>,
    <Array [1, 2, 3, 1.1, 2.2, 3.3] type='6 * union[int64[parameters={"__array__": "...'>,
]
```

The unique combos are `(0, 0)` and `(1, 1)` because the union-tags of `c` and `c` line up. All of the combos are `(0, 0)`, `(0, 1)`, `(1, 0)`, and `(1, 1)`, including two cases that never happen in these arrays. The old code iterated through the unique combos, recursively descending on each, producing two `contents` in the output UnionArray. The new code iterates through all combos, recursively descending on each (though some of these descents are through empty arrays, after masking the combo with an array of all combos), producing four `contents` in the output UnionArray. (I just checked, and they get `simplified` away, but the fact that it recursively descended into those empty arrays passes the type through to the final arrays.)

The bit I added yesterday in PR #808 is gone now, since the `numoutputs is None` case is now impossible. To get to this `if` branch, there must be at least one UnionArray, and every UnionArray must have at least one content, so the `for` loop over `all_combos` must have at least one iteration.
